botorch/posteriors/ensemble.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""
+Ensemble posteriors. Used in conjunction with ensemble models.
+"""
+
+from __future__ import annotations
+
+from typing import Optional
+
+import torch
+from botorch.posteriors.posterior import Posterior
+from torch import Tensor
+
+r"""
+Posterior module to be used with ensemble models.
+"""
+
+
+class EnsemblePosterior(Posterior):
+    r"""Ensemble posterior, that should be used for ensemble models that compute
+    eagerly a finite number of samples per X value as for example a deep ensemble
+    or a random forest."""
+
+    def __init__(self, values: Tensor) -> None:
+        r"""
+        Args:
+            values: Values of the samples produced by this posterior as
+                a `(b) x s x q x m` tensor where `m` is the output size of the
+                model and `s` is the ensemble size.
+        """
+        if values.ndim < 3:
+            raise ValueError("Values has to be at least three-dimensional.")
+        self.values = values
+
+    @property
+    def size(self) -> int:
+        r"""The size of the ensemble"""
+        return self.values.shape[-3]
+
+    @property
+    def weights(self) -> Tensor:
+        r"""The weights of the individual models in the ensemble.
+        Equally weighted by default."""
+        return torch.ones(self.size) / self.size
+
+    @property
+    def device(self) -> torch.device:
+        r"""The torch device of the posterior."""
+        return self.values.device
+
+    @property
+    def dtype(self) -> torch.dtype:
+        r"""The torch dtype of the posterior."""
+        return self.values.dtype
+
+    @property
+    def mean(self) -> Tensor:
+        r"""The mean of the posterior as a `(b) x n x m`-dim Tensor."""
+        return self.values.mean(dim=-3)
+
+    @property
+    def variance(self) -> Tensor:
+        r"""The variance of the posterior as a `(b) x n x m`-dim Tensor.
+
+        Computed as the sample variance across the ensemble outputs.
+        """
+        if self.size == 1:
+            return torch.zeros_like(self.values.squeeze(-3))
+        return self.values.var(dim=-3)
+
+    def _extended_shape(
+        self, sample_shape: torch.Size = torch.Size()  # noqa: B008
+    ) -> torch.Size:
+        r"""Returns the shape of the samples produced by the posterior with
+        the given `sample_shape`.
+        """
+        mask = self.values.ndim * [True]
+        mask[-3] = False
+        return sample_shape + tuple(torch.tensor(self.values.shape)[mask].tolist())
```

Comment:
```suggestion
        return sample_shape + self.values.shape[:1] + self.values.shape[-2:]
```

Reply:
This will not work, as the batch dim is optional. Consider first this case:
```
self.values = torch.randn(4,16,5,2)
```
Here the batch size is 4 and the ensemble size is 16. In the second part of the sum above we want to filter out the ensemble dimension and `self.values.shape[:1] + self.values.shape[-2:]` would correctly return `torch.Size([4, 5, 2])`.

But now consider the case that no batch dimension is provided which is allowed:
```
self.values = torch.randn(16,5,2)
```
Here again the ensemble size is 16, but the expression from above would return `torch.Size([16, 5, 2])` which is wrong, it should return `torch.Size([5, 2])`. For this reason, I went for the rather complicated solution from above and also originally preferred having the ensemble size as last dimension, but this is for other reasons not preferable.

Reply:
Got it, but `self.values.shape[:-3] + self.values.shape[-2:]` should work, right?

Reply:
Yes, thx for the hint. "Manchmal sieht man den Wald vor lauter Bäumen nicht" ;)

Reply:
Yeah the classic "sometimes see man the forest before noisy trees not"